### PR TITLE
Helpers to parse Foorm forms and submissions for analytics

### PR DESCRIPTION
**Pull request by bencodeorg**
Introduces new helpers, Foorm::FormAnalyticsParser and Foorm::SubmissionAnalyticsParser, which reshape Forms and Submissions, respectively, into a format for export in a tabular format.

Note this code isn't actually called anywhere right now -- if you were to call it (say, from the Rails console), it would write two CSVs to the Dashboard directory.

The structure of these helpers is similar to the existing Foorm::FoormParser helper, with a bunch of methods defined on a class. I'm not sure if this is a great practice (there's nothing really "object-y" about this), so would love input on what format makes sense (a module?)

## Testing story

I've added unit tests to cover the types of questions that currently appear in Forms. I also manually tested with the limited set of Forms and Submissions that I have in my local database.

## Follow-up work

- I want to test this on a production clone and see how it performs, from a memory utilization / database load / timing perspective.
- We will likely need to implement some batching here, such that all Submissions are not loaded into memory at once.
- Suresh has suggested his preference would be that we do not put these in a CSV in S3 as an intermediate step -- in any case, we'll have to come up with some solution for actually loading this data into Redshift.
- There are more steps to this work (setting up cron to run regularly, etc.), but I won't list them all here.

## PR Checklist:
- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/models/foorm/submission.rb [hunk 1]
```diff
@@ -43,25 +43,25 @@ def formatted_answers
     parsed_answers = JSON.parse(answers)
 
     # Then, merge in formatted versions of each answer in the submission.
-    parsed_answers.each do |question_id, answer|
-      formatted_answers.merge! get_formatted_answer(question_id, answer)
+    parsed_answers.each do |question_name, answer|
```
Comment (bencodeorg) [inline dashboard/app/models/foorm/submission.rb:46]: These changes are unrelated cleanup, to better suggest that the keys in `parsed_answers` are the `name` field for each `question` in a Form.

File: dashboard/lib/pd/foorm/form_analytics_parser.rb [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+          reshaped_form.each do |question|
+            row = question.stringify_keys.values_at(*HEADERS)
+            csv << row
+          end
+        end
+      end
+    end
+
+    def self.reshape_form(form)
```
Comment (bencodeorg) [inline dashboard/lib/pd/foorm/form_analytics_parser.rb:45]: Maybe this would make more sense as a mix-in for (instances of) the Form class?

File: dashboard/lib/pd/foorm/form_analytics_parser.rb [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+      is_facilitator_specific
+      response_options
+      num_response_options
+    )
+
+    def self.reshape_all_forms_and_export_to_csv
+      CSV.open('./test_forms.csv', 'wb') do |csv|
+        csv << HEADERS
+
+        # Loads 1000 records at a time to manage loading records into memory.
```
Comment (jamescodeorg) [inline dashboard/lib/pd/foorm/form_analytics_parser.rb:30]: I don't think the batching is implemented yet, right?  Maybe change this comment into a TODO?

File: dashboard/lib/pd/foorm/form_analytics_parser.rb [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+      question_name
+      question_text
+      matrix_item_name
+      matrix_item_text
+      is_facilitator_specific
+      response_options
+      num_response_options
+    )
+
+    def self.reshape_all_forms_and_export_to_csv
```
Comment (jamescodeorg) [inline dashboard/lib/pd/foorm/form_analytics_parser.rb:26]: For each of these "reshape" methods, it would be helpful to have a header comment that briefly describes the input and output formats.

File: dashboard/lib/pd/foorm/form_analytics_parser.rb [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+      CSV.open('./test_forms.csv', 'wb') do |csv|
```
Comment (jamescodeorg) [inline dashboard/lib/pd/foorm/form_analytics_parser.rb:27]: Is this the name that you want for the csv file?

File: dashboard/lib/pd/foorm/submission_analytics_parser.rb [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    def self.reshape_submission_answers(parsed_answers, form)
+      reshaped_submission_answers = []
+
+      parsed_answers.each do |question_name, answer|
+        reshaped_submission_answer = {question_name: question_name}
+        question_details = form&.get_question_details(question_name)
+
+        # If question isn't in the Form, return as-is.
+        # This is expected for metadata about the submission,
+        # which doesn't have an associated question in the Form.
```
Comment (jamescodeorg) [inline dashboard/lib/pd/foorm/submission_analytics_parser.rb:64]: This seems a bit kludgy.  For my education, where is the extra metadata inserted?

File: dashboard/lib/pd/foorm/submission_analytics_parser.rb [hunk 7]
```diff
@@ -0,0 +1,112 @@
+module Pd::Foorm
```
Comment (jamescodeorg) [inline dashboard/lib/pd/foorm/submission_analytics_parser.rb:1]: (All comments for the other parser also apply to this one.)

**Review by jamescodeorg — COMMENTED** [review]
General structure looks good, a few questions/comments below.

File: dashboard/lib/pd/foorm/form_analytics_parser.rb (hunk 2 above)
Comment (jamescodeorg) [reply dashboard/lib/pd/foorm/form_analytics_parser.rb:45]: (I'm sorry, I never really understood how to use mixins so I don't know the answer to this question without doing more research.)

File: dashboard/lib/pd/foorm/form_analytics_parser.rb (hunk 5 above)
Comment (bencodeorg) [reply dashboard/lib/pd/foorm/form_analytics_parser.rb:27]: I still need to iron out how/where the CSV export will happen, so these are just placeholders.

File: dashboard/lib/pd/foorm/form_analytics_parser.rb (hunk 3 above)
Comment (bencodeorg) [reply dashboard/lib/pd/foorm/form_analytics_parser.rb:30]: I updated to use basic batching this morning -- `find_each` works in batches (vs `all`, which does not).

File: dashboard/lib/pd/foorm/submission_analytics_parser.rb (hunk 6 above)
Comment (bencodeorg) [reply dashboard/lib/pd/foorm/submission_analytics_parser.rb:64]: I believe on save of the Submission (ie, when the user completes a Form) to the DB. We include information like a variable containing the course about which the user completed the survey (eg, CS Principles vs. CS Discoveries), or other variables that influence what questions are shown / are substituted into question text.

File: dashboard/lib/pd/foorm/form_analytics_parser.rb (hunk 3 above)
Comment (jamescodeorg) [reply dashboard/lib/pd/foorm/form_analytics_parser.rb:30]: Ah, I see.  `find_each` doesn't really say "batched" to me so I might consider explicitly passing `batch_size` as a named parameter to clarify this.  Up to you.

Comment (bencodeorg) [reply dashboard/lib/pd/foorm/form_analytics_parser.rb:30]: Yeah I added the comment because I agree, but I think documenting in code is even better, thanks!

**Review by jamescodeorg — APPROVED** [review]
